drake/solvers/Optimization.h

```diff
@@ -114,6 +114,8 @@ namespace Drake {
     }
     virtual ~Constraint() {}
 
+    // TODO: consider using a Ref for `y` too.  This will require the client
+    // to do allocation, but also allows it to choose stack allocation instead.
     virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
     virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const = 0;  // move this to DifferentiableConstraint derived class if/when we need to support non-differentiable functions
```

Comment:
Yes.  That will have to be worked out as part of the functional form work.  In the current implementation the vtable of `ConstraintImpl<F>` will force a call to both overloads of `F::eval` to be compiled.  In order to avoid that we will have to learn through `FunctionTraits<F>` somehow whether each overload is expected.